Feat: Adding poll support

budgetdevv opened the pull request:
Add poll support to NetCord.

budgetdevv committed:
Added data-structures for REST API

budgetdevv commented:
It is worth noting that this is still WIP. Made a PR just so progress can be tracked

budgetdevv committed:
Added MessagePollAnswer.Id

budgetdevv committed:
Added MessagePollAnswerCount.cs and MessagePollResults.cs

budgetdevv committed:
Avoid hiding JsonEntity.Id, renamed to AnswerId

budgetdevv committed:
Made some properties `required`

KubaZ2 commented:
NetCord doesn't use `required` keyword anywhere currently. It would be nice not to break the rule. I would make constructors with required/common values instead. Also I don't think making a poll base class is a good idea. Mixing send and receive properties is not what NetCord does currently.

budgetdevv commented:
Okay, will try to refactor with what you said in mind

KubaZ2 commented on NetCord/Rest/MessagePollAnswerCount.cs:
```diff
@@ -0,0 +1,17 @@
+
+public partial class MessagePollAnswerCount : JsonEntity
+{
+    [JsonPropertyName("id")]
+    public int AnswerId { get; set; }
```
It's a duplicate of the `Id` property defined in `JsonEntity` and I think it should also be an `ulong`.

budgetdevv replied:
Yup already changed that in my local commit

budgetdevv commented:
I also realized that I have to make IJsonModel wrappers for non-properties ( We discussed this in Discord )

KubaZ2 commented on NetCord/Rest/MessagePoll.cs:
```diff
@@ -0,0 +1,10 @@
+
+public partial class MessagePoll : MessagePollBase
+{
+    // Non-expiring posts are possible in the future, see: https://github.com/discord/discord-api-docs/blob/e4bdf50f11f9ca61ace2636285e029a2b3dfd0ec/docs/resources/Poll.md#poll-object
+    [JsonPropertyName("expiry")]
+    public DateTimeOffset? Expiry { get; set; }
```
It should be named `ExpiresAt` to match the current code naming I think.

budgetdevv committed:
- Removed MessagePollBase.cs ( Kuba said send / receive shouldn't mix )

- Removed usages required keyword ( Kuba said that this will be a breaking change )

- Renamed Expiry to ExpireAt for consistency

- Use JsonEmoji instead of Emoji in json data-structures

- Made wrapper entities

KubaZ2 commented on NetCord/Rest/JsonModels/JsonMessagePoll.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace NetCord.Rest.JsonModels;
+
+public partial class JsonMessagePoll: JsonEntity
```
```suggestion
public partial class JsonMessagePoll : JsonEntity
```

KubaZ2 commented on NetCord/Rest/JsonModels/JsonMessagePoll.cs:
```diff
@@ -0,0 +1,24 @@
+public partial class JsonMessagePoll: JsonEntity
+{
+    [JsonPropertyName("question")]
+    public JsonMessagePollMedia Question { get; set; } = null!;
```
```suggestion
    public JsonMessagePollMedia Question { get; set; }
```

KubaZ2 commented on NetCord/Rest/JsonModels/JsonMessagePoll.cs:
```diff
@@ -0,0 +1,24 @@
+    public JsonMessagePollMedia Question { get; set; } = null!;
+    
+    [JsonPropertyName("answers")]
+    public JsonMessagePollAnswer[] Answers { get; set; } = null!;
```
```suggestion
    public JsonMessagePollAnswer[] Answers { get; set; }
```

budgetdevv committed:
Get rid of required keyword

budgetdevv committed:
- Added JsonMessagePollAnswerCount.cs
- Fixed type of answer_id

budgetdevv committed:
Get rid of null! usage

budgetdevv committed:
Fixed formatting

Co-authored-by: Kuba_Z2 <77853483+KubaZ2@users.noreply.github.com>

budgetdevv committed:
- Remove required keyword, suppress using null! ( Rider seems to complain about it, and it caused checks to fail earlier on )

- Fixed properties

budgetdevv committed:
- Added MessagePoll.Results
- Namespace fix

budgetdevv committed:
Don't use null!

budgetdevv committed:
Added MessagePollMediaProperties.cs and MessagePollAnswerProperties.cs

budgetdevv committed:
Added constructor to MessagePollAnswerProperties.cs

budgetdevv committed:
Optimize namespace

budgetdevv committed:
Added comments

budgetdevv commented:
Finally checks are passing

budgetdevv committed:
Added RestClient.Undocumented.ExpirePollAsync(). Not too sure if it works since my understanding of [GenerateAlias] is limited

KubaZ2 commented on NetCord/Rest/RestClient.Undocumented.cs:
```diff
@@ -40,4 +40,13 @@ public IAsyncEnumerable<GuildUserInfo> SearchGuildUsersAsync(ulong guildId, Guil
+    
+    [GenerateAlias(typeof(RestMessage), nameof(RestMessage.Id))]
+    [GenerateAlias(typeof(TextChannel), nameof(TextChannel.Id))]
+    public async Task<RestMessage> ExpirePollAsync(ulong messageId, ulong channelId, RestRequestProperties? properties = null)
```
```suggestion
    public async Task<RestMessage> ExpirePollAsync(ulong channelId, ulong messageId, RestRequestProperties? properties = null)
```

KubaZ2 commented on NetCord/Rest/RestClient.Undocumented.cs:
```diff
@@ -40,4 +40,13 @@ public IAsyncEnumerable<GuildUserInfo> SearchGuildUsersAsync(ulong guildId, Guil
             properties);
     }
+    
+    [GenerateAlias(typeof(RestMessage), nameof(RestMessage.Id))]
```
```suggestion
    [GenerateAlias(typeof(RestMessage), nameof(RestMessage.ChannelId), nameof(RestMessage.Id))]
```

KubaZ2 commented on NetCord/MessagePollResults.cs:
```diff
@@ -0,0 +1,18 @@
+
+public partial class MessagePollResults : IJsonModel<JsonMessagePollResults>
+{
+    public JsonMessagePollResults JsonModel { get; }
```
`IJsonModel<T>` should be implemented explicitly EVERYWHERE

KubaZ2 commented on NetCord/MessagePollProperties.cs:
```diff
@@ -0,0 +1,27 @@
+    [JsonPropertyName("answers")]
+    public MessagePollAnswerProperties[] Answers { get; set; }
+
+    public MessagePollProperties(MessagePollMediaProperties question, MessagePollAnswerProperties[] answers)
+    {
+        Question = question;
+        Answers = answers;
+    }
```
You can use primary constructor here, also the value of answers and the corresponding property should be `IEnumerable<MessagePollAnswerProperties>`

KubaZ2 commented on NetCord/MessagePollAnswer.cs:
```diff
@@ -0,0 +1,21 @@
+    public ulong AnswerId { get; }
+    public MessagePollMedia PollMedia { get; }
+
+    [SetsRequiredMembers]
```
```suggestion
```

KubaZ2 commented on NetCord/MessagePoll.cs:
```diff
@@ -0,0 +1,27 @@
+public partial class MessagePoll : IJsonModel<JsonMessagePoll>
+{
+    public JsonMessagePoll JsonModel { get; }
+    public MessagePollMedia Question { get; set; } = null!;
```
```suggestion
    public MessagePollMedia Question { get; set; }
```

KubaZ2 commented on NetCord/MessagePollAnswerCount.cs:
```diff
@@ -0,0 +1,11 @@
+public partial class MessagePollAnswerCount(JsonMessagePollAnswerCount jsonModel) : IJsonModel<JsonMessagePollAnswerCount>
+{
+    public JsonMessagePollAnswerCount JsonModel { get; } = jsonModel;
+    public ulong AnswerId { get; } = jsonModel.Id;
+    public int Count { get; } = jsonModel.Count;
+    public bool MeVoted { get; } = jsonModel.MeVoted;
```
The properties should call the json model instead of storing the values

KubaZ2 commented on NetCord/JsonModels/JsonMessagePollResults.cs:
```diff
@@ -0,0 +1,14 @@
+    [JsonPropertyName("answer_counts")]
+    public MessagePollAnswerCount[]? Answers { get; set; }
+
+    public bool ContainsAnswers => Answers != null;
```
I don't think we want properties like that in json models.

budgetdevv committed:
Removed property in JsonMessagePollResults.cs

budgetdevv committed:
Removed initializer

budgetdevv committed:
Removed MessagePoll's setters

budgetdevv committed:
Read values from JsonModel instead of storing them

budgetdevv committed:
Removed SetsRequiredMembers attribute

budgetdevv committed:
- Convert more constructors to primary ones.
- Implement IJsonModel.JsonModel explicitly.

budgetdevv committed:
MessagePoll.cs:
- Fixed inproper initialization.
- Use IReadOnlyList over array

budgetdevv committed:
Prefer IEnumerable over array.

KubaZ2 reviewed (commented):
Why are most classes partial? Only options and properties must be partial.

budgetdevv commented:
Alright, I will make further changes with that in mind

budgetdevv committed:
Removed redundant partial keyword

KubaZ2 commented on NetCord/MessagePoll.cs:
```diff
@@ -0,0 +1,18 @@
+{
+    JsonMessagePoll IJsonModel<JsonMessagePoll>.JsonModel => jsonModel;
+    public MessagePollMedia Question = new(jsonModel.Question, guildId, client);
+    public IReadOnlyList<MessagePollAnswer> Answers = jsonModel.Answers.Select(x => new MessagePollAnswer(x, guildId, client)).ToImmutableList();
```
```suggestion
    public IReadOnlyList<MessagePollAnswer> Answers = jsonModel.Answers.Select(x => new MessagePollAnswer(x, guildId, client)).ToArray();
```

KubaZ2 commented on NetCord/MessagePollResults.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    JsonMessagePollResults IJsonModel<JsonMessagePollResults>.JsonModel => jsonModel;
+    public bool IsFinalized => jsonModel.IsFinalized;
+    public MessagePollAnswerCount[]? Answers => jsonModel.Answers;
```
```suggestion
    public IReadOnlyList<MessagePollAnswerCount>? Answers => jsonModel.Answers;
```

KubaZ2 commented on NetCord/JsonModels/JsonMessagePollResults.cs:
```diff
@@ -0,0 +1,12 @@
+    public bool IsFinalized { get; set; }
+    
+    [JsonPropertyName("answer_counts")]
+    public MessagePollAnswerCount[]? Answers { get; set; }
```
A JSON model should be used here I think

budgetdevv committed:
Revert change to IReadOnlyList

Co-authored-by: Kuba_Z2 <77853483+KubaZ2@users.noreply.github.com>

budgetdevv committed:
Use IReadOnlyList

Co-authored-by: Kuba_Z2 <77853483+KubaZ2@users.noreply.github.com>

budgetdevv committed:
Fixed JsonMessagePollResults.cs

budgetdevv committed:
Revert back to IReadOnlyList for MessagePollResults.Answers, it seems like I misunderstood what Kuba was trying to tell me

KubaZ2 commented on NetCord/MessagePollResults.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    JsonMessagePollResults IJsonModel<JsonMessagePollResults>.JsonModel => jsonModel;
+    public bool IsFinalized => jsonModel.IsFinalized;
+    public IReadOnlyList<MessagePollAnswerCount>? Answers = jsonModel.Answers?.Select(x => new MessagePollAnswerCount(x)).ToArray();
```
I think a normal constructor with an `if` statement should be used instead of the `?`, that is what NetCord does to prevent setting null to null unnecessarily.

KubaZ2 commented on NetCord/JsonModels/JsonMessagePoll.cs:
```diff
@@ -0,0 +1,25 @@
+    
+    // Non-expiring posts are possible in the future, see: https://github.com/discord/discord-api-docs/blob/e4bdf50f11f9ca61ace2636285e029a2b3dfd0ec/docs/resources/Poll.md#poll-object
+    [JsonPropertyName("expiry")]
+    public DateTimeOffset? ExpireAt { get; set; }
```
```suggestion
    public DateTimeOffset? ExpiresAt { get; set; }
```

KubaZ2 commented on NetCord/MessagePoll.cs:
```diff
@@ -0,0 +1,18 @@
+    public bool AllowMultiselect => jsonModel.AllowMultiselect;
+    public MessagePollLayoutType LayoutType => jsonModel.LayoutType;
+    // Non-expiring posts are possible in the future, see: https://github.com/discord/discord-api-docs/blob/e4bdf50f11f9ca61ace2636285e029a2b3dfd0ec/docs/resources/Poll.md#poll-object
+    public DateTimeOffset? ExpireAt => jsonModel.ExpireAt;
```
```suggestion
    public DateTimeOffset? ExpiresAt => jsonModel.ExpiresAt;
```

KubaZ2 commented on NetCord/JsonModels/JsonMessagePollAnswerCount.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace NetCord.JsonModels;
+
+public class JsonMessagePollAnswerCount : JsonEntity
```
The id for the object is documented as an integer, so it shouldn't be a JsonEntity and the `Id` property should be an int I think. The same rule applies to other objects too.

budgetdevv committed:
Made AnswerId a uint

budgetdevv committed:
Renamed ExpireAt to ExpiresAt for consistency

budgetdevv committed:
AnswerId is now a uint for MessagePollAnswer.cs ( And its JSON variant )

budgetdevv committed:
It seems like MessagePollResults.Answers is is not nullable. An empty list / array is probably sent should there be no votes. TODO: Verify this assumption

budgetdevv committed:
AnswerId should be uint

budgetdevv committed:
It seems like this is only sent via gateway, commenting out for now...

budgetdevv committed:
Fixed RestMessage.ExpirePollAsync() ( Hopefully )

budgetdevv committed:
Make MessagePollProperties.LayoutType nullable.

budgetdevv committed:
Re-included MessagePollResults.ContainsAnswers

KubaZ2 commented:
Why is the AnwerId an uint now? Isn't it better to make it a signed integer? I personally would use unsigned types everywhere, but since .NET BCL doesn't use them (probably because they are not CLS compliant), I don't think properties like that should be unsigned unless there is some good reason for that. Also I don't think the AnswerId can be grater than or equal to the Count, but the Count is an int.

budgetdevv commented:
Because count can never be negative, also we use ulong for IDs

budgetdevv commented:
Which count are you referring to btw?

KubaZ2 commented:
`MessagePollAnswerCount.Count`

budgetdevv commented:
Yeah because collections use int for count

budgetdevv commented:
I do not think this is a concern however, NetCord is unlikely to be ported to another language. But if int is used throughout the project for count, then I should probably make it an int for the sake of consistency

KubaZ2 commented:
> Because count can never be negative, also we use ulong for IDs

Snowflakes are documented as e.g. a uint64 and using the unsigned type makes the lib work longer. But tbh changing it to a long wouldn't be a bad idea.

KubaZ2 commented:
> I do not think this is a concern however, NetCord is unlikely to be ported to another language. But if int is used throughout the project for count, then I should probably make it an int for the sake of consistency

NetCord is not CLS compliant so you are probably correct. Also I didn't mean I do care about the CLS compliance.

budgetdevv commented:
So should I change them to unsigned int instead?

KubaZ2 commented:
I would use signed integers.

budgetdevv committed:
Made ids signed integer

budgetdevv committed:
Optimize namespace

budgetdevv committed:
Added InteractionMessageProperties.Poll

budgetdevv committed:
- Implemented RestMessage.GetPollAnswerVoters() ( Need to figure out how to implement pagination support )

- RestMessage.ExpirePollAsync() now takes into account of RestRequestProperties

budgetdevv commented:
I realized that I'm returning the wrong type for .GetPollAnswerVoters(), will fix it

budgetdevv committed:
- Added JsonGetPollAnswerVotersResult.cs

- Updated Serialization.cs to generate for JsonGetPollAnswerVotersResult.cs

- Renamed RestClient.GetPollAnswerVoters() to RestClient.GetPollAnswerVotersAsync()

- Fixed erroneous return type for RestClient.GetPollAnswerVotersAsync()

budgetdevv committed:
Made afterUserId an optional parameter

budgetdevv committed:
Default DurationInHours to 1 ( Via optional ctor param )

budgetdevv commented:
However, it seems like after running ExpirePollAsync() ( Which works ), MessagePollResults.Answers is still an empty array, with IsFinalized set to false

budgetdevv committed:
Fixed HttpMethod for RestMessage.GetPollAnswerVotersAsync() ( How does one be so dumb??? )

budgetdevv committed:
Added convenience API for constructing MessagePollAnswerProperties.cs

budgetdevv committed:
Made params in primary ctor optional

budgetdevv committed:
Merge branch 'NetCordDev:alpha' into polls

budgetdevv committed:
Added MESSAGE_POLL_VOTE_ADD and MESSAGE_POLL_VOTE_REMOVE gateway events. Pending tests

budgetdevv committed:
Added poll intents, made intents less complicated

budgetdevv committed:
Added Poll field to WebhookMessageProperties.cs

budgetdevv committed:
Merge branch 'NetCordDev:alpha' into polls

budgetdevv committed:
Adapt [GenerateAlias] to new NetCord commit

budgetdevv committed:
Added Poll property to IPartialMessage / PartialMessage

budgetdevv committed:
Source generate ExpirePollAsync() and GetPollAnswerVotersAsync() for IPartialMessage

budgetdevv committed:
Message poll results are nullable

KubaZ2 committed:
Improvements

KubaZ2 committed:
Merge branch 'alpha' into polls

KubaZ2 committed:
Work

KubaZ2 committed:
Cleanup

KubaZ2 committed:
Delete a comment

KubaZ2 committed:
Add missing `GenerateAliasAttribute`s

KubaZ2 commented:
Thanks!

budgetdevv commented:
You're welcome for nothing, sorry I haven't had time to work on it ( School sucks )
